ns for fx: add support for shadowing linear fp16 patterns

vkuzo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #55155 ns for fx: rename SugraphTypeRelationship to SubgraphTypeRelationship
* #55154 ns for fx: add allowlist for ops with same signature across dtypes
* #55080 ns for fx: add linear-relu mod weight extraction
* #55079 ns for fx: weight extaction for conv1d and conv3d
* #55078 ns for fx: ensure kwargs are handled when graph matching
* #55077 ns for fx: clean up debug print statements
* #55060 ns for fx: move it to top level file
* #54335 ns for fx: weight extraction for nni.ConvReLU2d
* #54327 ns for fx: make shadow input logging work for multi node subgraphs
* #54326 ns for fx: make input logging work for multi-node subgraphs
* #54280 ns for fx: refactor test cases
* **#54275 ns for fx: add support for shadowing linear fp16 patterns**

Summary:

Adds support for NS shadow activations path for the fp16 emulation
pattern such as

```
... -> dequantize -> linear -> relu -> to(torch.float16) -> ...
```

There are a couple of changes necessary here:

1. removing the restriction on the shadowing graph pass that the B
subgraph is a single node (since this subgraph is four nodes), and
modifying the code to correctly add the relevant inputs versus output
loggers (input loggers and subgraph copy if we are at start_node,
and output logger if we are at end_node)

2. modifying the logic for calculating node input and output type
to work correcty for the `to` and `dequantize` nodes:
2a. make the function return the first input and output, instead of just
the first input
2b. make the function handle `dequantize` correctly by recursively
using the output if its input
2c. make the function handle `to` correctyl by recursively using the
output of its input and the target dtype

3. a bug fix to handle observers in kwargs, while copying subgraphs

Note: input logging for these patterns is not tested yet,
this will be in the next PR.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_linear_fp16
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27172655](https://our.internmc.facebook.com/intern/diff/D27172655)

vkuzo committed:
ns for fx: add support for shadowing linear fp16 patterns

Summary:

Adds support for NS shadow activations path for the fp16 emulation
pattern such as

```
... -> dequantize -> linear -> relu -> to(torch.float16) -> ...
```

There are a couple of changes necessary here:

1. removing the restriction on the shadowing graph pass that the B
subgraph is a single node (since this subgraph is four nodes), and
modifying the code to correctly add the relevant inputs versus output
loggers (input loggers and subgraph copy if we are at start_node,
and output logger if we are at end_node)

2. modifying the logic for calculating node input and output type
to work correcty for the `to` and `dequantize` nodes:
2a. make the function return the first input and output, instead of just
the first input
2b. make the function handle `dequantize` correctly by recursively
using the output if its input
2c. make the function handle `to` correctyl by recursively using the
output of its input and the target dtype

3. a bug fix to handle observers in kwargs, while copying subgraphs

Note: input logging for these patterns is not tested yet,
this will be in the next PR.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_linear_fp16
```

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vkuzo committed:
Update on "ns for fx: add support for shadowing linear fp16 patterns"


Summary:

Adds support for NS shadow activations path for the fp16 emulation
pattern such as

```
... -> dequantize -> linear -> relu -> to(torch.float16) -> ...
```

There are a couple of changes necessary here:

1. removing the restriction on the shadowing graph pass that the B
subgraph is a single node (since this subgraph is four nodes), and
modifying the code to correctly add the relevant inputs versus output
loggers (input loggers and subgraph copy if we are at start_node,
and output logger if we are at end_node)

2. modifying the logic for calculating node input and output type
to work correcty for the `to` and `dequantize` nodes:
2a. make the function return the first input and output, instead of just
the first input
2b. make the function handle `dequantize` correctly by recursively
using the output if its input
2c. make the function handle `to` correctyl by recursively using the
output of its input and the target dtype

3. a bug fix to handle observers in kwargs, while copying subgraphs

Note: input logging for these patterns is not tested yet,
this will be in the next PR.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_linear_fp16
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27172655](https://our.internmc.facebook.com/intern/diff/D27172655)

[ghstack-poisoned]

vkuzo committed:
Update on "ns for fx: add support for shadowing linear fp16 patterns"


Summary:

Adds support for NS shadow activations path for the fp16 emulation
pattern such as

```
... -> dequantize -> linear -> relu -> to(torch.float16) -> ...
```

There are a couple of changes necessary here:

1. removing the restriction on the shadowing graph pass that the B
subgraph is a single node (since this subgraph is four nodes), and
modifying the code to correctly add the relevant inputs versus output
loggers (input loggers and subgraph copy if we are at start_node,
and output logger if we are at end_node)

2. modifying the logic for calculating node input and output type
to work correcty for the `to` and `dequantize` nodes:
2a. make the function return the first input and output, instead of just
the first input
2b. make the function handle `dequantize` correctly by recursively
using the output if its input
2c. make the function handle `to` correctyl by recursively using the
output of its input and the target dtype

3. a bug fix to handle observers in kwargs, while copying subgraphs

Note: input logging for these patterns is not tested yet,
this will be in the next PR.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_linear_fp16
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27172655](https://our.internmc.facebook.com/intern/diff/D27172655)

[ghstack-poisoned]

vkuzo committed:
Update on "ns for fx: add support for shadowing linear fp16 patterns"


Summary:

Adds support for NS shadow activations path for the fp16 emulation
pattern such as

```
... -> dequantize -> linear -> relu -> to(torch.float16) -> ...
```

There are a couple of changes necessary here:

1. removing the restriction on the shadowing graph pass that the B
subgraph is a single node (since this subgraph is four nodes), and
modifying the code to correctly add the relevant inputs versus output
loggers (input loggers and subgraph copy if we are at start_node,
and output logger if we are at end_node)

2. modifying the logic for calculating node input and output type
to work correcty for the `to` and `dequantize` nodes:
2a. make the function return the first input and output, instead of just
the first input
2b. make the function handle `dequantize` correctly by recursively
using the output if its input
2c. make the function handle `to` correctyl by recursively using the
output of its input and the target dtype

3. a bug fix to handle observers in kwargs, while copying subgraphs

Note: input logging for these patterns is not tested yet,
this will be in the next PR.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_linear_fp16
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27172655](https://our.internmc.facebook.com/intern/diff/D27172655)

[ghstack-poisoned]

hx89 commented on torch/quantization/ns/graph_passes.py:
```diff
@@ -154,12 +154,19 @@ def _insert_dtype_cast_after_node(
+        (node_input_type_a == NodeInputOrOutputType.FP32 and
+         node_input_type_c == NodeInputOrOutputType.INT8) or
+        (node_input_type_a == NodeInputOrOutputType.FP32 and
+         node_input_type_c == NodeInputOrOutputType.FP16)
```
A general question, is this change for supporting Fakelowp ops? IIRC the output of Fakelowp ops is fp32 but casted to fp16 precision, so I guess it's for something else?

vkuzo replied:
yes, this is for fakelowp ops.  For example,

```
# original graph
x0 -> linear_fp32 -> x1

# graph with fakelowp replacement
x0_fp16 -> dequantize -> linear_fp32 -> fp32_to_fp16 -> x1_fp16

# graph with fp32 linear shadowing fakelowp linear
  dequantize -> linear_fp32 -> logger
    /
x0_fp16 -> dequantize -> linear_fp32 -> fp32_to_fp16 -> logger -> x1_fp16
```

vkuzo committed:
Update on "ns for fx: add support for shadowing linear fp16 patterns"


Summary:

Adds support for NS shadow activations path for the fp16 emulation
pattern such as

```
... -> dequantize -> linear -> relu -> to(torch.float16) -> ...
```

There are a couple of changes necessary here:

1. removing the restriction on the shadowing graph pass that the B
subgraph is a single node (since this subgraph is four nodes), and
modifying the code to correctly add the relevant inputs versus output
loggers (input loggers and subgraph copy if we are at start_node,
and output logger if we are at end_node)

2. modifying the logic for calculating node input and output type
to work correcty for the `to` and `dequantize` nodes:
2a. make the function return the first input and output, instead of just
the first input
2b. make the function handle `dequantize` correctly by recursively
using the output if its input
2c. make the function handle `to` correctyl by recursively using the
output of its input and the target dtype

3. a bug fix to handle observers in kwargs, while copying subgraphs

Note: input logging for these patterns is not tested yet,
this will be in the next PR.

Test Plan:

```
python test/test_quantization.py TestFXNumericSuiteCoreAPIs.test_linear_fp16
```

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D27172655](https://our.internmc.facebook.com/intern/diff/D27172655)

[ghstack-poisoned]
